docs/adr/0002-use-react-redux-and-typescript-for-subscription-management-pages.md

```diff
@@ -0,0 +1,129 @@
+encapsulating reusable UI functionality in "hooks" and JS classes.
+
+[React]: https://reactjs.org/
+
+[Redux][] is a library for managing app state in a singleton container using a
+defined inventory of actions that describe changes to be applied via reducer
+functions and middleware filters.
+
+[Redux]: https://redux.js.org/
+
+[TypeScript][] is a superset of JavaScript which introduces type annotation
+expressions and type inference along with tooling to lint and compile.
+
+[TypeScript]: https://www.typescriptlang.org/
+
+We can also consider other dependencies that act as glue between these pieces
+and utilities to reduce boilerplate and repetition. (e.g. [redux-actions][] to
+simplify constructing actions; [redux-promise-middleware][] for actions that
+result in async API fetches; [redux-thunk][] for producing action sequences)
+
+[redux-actions]: https://redux-actions.js.org/ 
+
+[redux-promise-middleware]:
+https://github.com/pburtchaell/redux-promise-middleware 
+
+[redux-thunk]: https://github.com/reduxjs/redux-thunk
+
+* Pros
+  * Frequently updated and vibrant ecosystem
+  * New team members may be more familiar with the React / Redux / TypeScript
+    ecosystem
+  * React integrates well with Webpack for code-splitting and lazy-loaded
+    component JS bundles
+  * React and Webpack can support per-component hot reloading rather than
+    reloading the whole page
+  * React can organize UI rendering into small components with single
+    responsibility for each
+  * React hooks can be used to encapsulate reusable chunks of UI logic & state
+    between components
+  * Redux can organize business logic and API requests in a central set of
+    modules
+  * Data-driven React components can be loosely-coupled with Redux state
+  * The Redux dev tools browser add-on can monitor, inspect, and even rewind app
+    state changes
+  * TypeScript can help document & validate data structures at build time (e.g.
+    in Redux state) and integrate with IDEs (e.g. intellisense hints in VS Code)
+  * Can follow the Mocha / Chai / Sinon style unit testing stack used elsewhere
+    in the project
+* Cons
+  * Novel technology stack for the project in general.
+  * Fast-moving tech stack that may need chasing after.
+  * Reduced code sharing with the rest of the project, reinventing some wheels.
+
+### Backbone.js
+
+[Backbone.js][] is small framework that offers base View and Model abstractions.
+
+[Backbone.js]: https://backbonejs.org/
+
+FxA combines it with a few other utilities including [Mustache][] for HTML
+templates; [Cocktail][] for class mixins; [jQuery][] for DOM manipulation.
+Depends on [Underscore.js][] for general data manipulation.
+
+[Mustache]: https://mustache.github.io/ 
+
+[Cocktail]: https://github.com/onsi/cocktail
+
+[jQuery]: https://jquery.com/
+
+[Underscore.js]: https://underscorejs.org/
+
+* Pros
+  * Framework of record for client-side code in FxA so far.
```

Comment:
There are other reasons to use Backbone that I'm missing. It's been a long while since I was in the thick of using it. Could help to get some battle-tested reasons added here.

Reply:
I'm not very familiar with React, but is there a general performance improvement vs BackboneJS? Maybe through code optimization or tooling etc?

One thing I like about Backbone is that the views can be reused for different purposes (although React might offer the same benefit?)

Reply:
> I'm not very familiar with React, but is there a general performance improvement vs BackboneJS? Maybe through code optimization or tooling etc?

I'd say React trades a bit of raw performance for convenience. The main trade off is moving the responsibility for thinking about DOM manipulations from the developer's head to an algorithm on the client. 

So, rather than thinking about changing individual DOM nodes or attributes, you write a function that returns a DOM tree from scratch based on current component properties. The virtual DOM algorithm compares and does a diff/patch to change the real DOM to match. There's also a lot of optimization to skip some or all of the process depending on whether component properties or output has changed.

> One thing I like about Backbone is that the views can be reused for different purposes (although React might offer the same benefit?)

In React, components can basically be functions that take in an object and output (virtual) DOM as a result of the object properties. So, you can extract reusable little views something like this:
```javascript
const User = ({ displayName, avatar }) =>
  <div className="profile"><img src={avatar} /> {displayName}</div>;
```
Elsewhere, you can use this like so:
```javascript
<div className="users">
  {users.map(({ displayName, avatar }) =>
    <User displayName={displayName} avatar={avatar} /> }
</div>
```
This is super-simple, but basically anything you can do with functions you can do with React components - e.g. composing functions, wrapping elements, transforming props, etc.

Reply:
Judging by a [few](https://medium.com/preply-engineering/migrating-from-backbone-js-to-react-js-without-losing-it-a8b383a102fe) [migration notes](https://www.clayallsopp.com/posts/from-backbone-to-react/), there's some performance wins for React and better re-use in theory. The first link has more detailed information on how to migrate an existing Backbone app at a later time if that becomes appealing.

Reply:
I'd say one way React actually does open up opportunities for performance is that it dramatically simplifies how you deal with DOM changes on the end-developer side, centrallizing the whole concern in an algorithm that has a ton of smart people banging on it in interesting ways.

Reply:
Also, FWIW, I didn't like HTML-in-JavaScript (i.e. JSX) when I first saw it. But, I felt better when I realized it's basically semantic sugar that gets transpiled down to something simple like:
```javascript
const h = React.createElement;
const User = ({ displayName, avatar }) =>
  h('div', { className: "profile" }, h('img', { src: avatar }), displayName);
```

Reply:
@lmorchard Thanks for examples and explanation, that clears things up for me 👍

Reply:
The good thing is that @hritvi is going to be doing some Backbone->React migration exploration in the settings page for the summer, so we'll see how that experience compares to the linked notes.

Reply:
FWIW, I'm *kinda* hoping some of what I'm doing in this PR and for #961 could help lay some groundwork for that effort this summer.